### PR TITLE
Coco cat training and evaluation

### PR DESCRIPTION
**Pull request by BlairLee**
# Peer Review Information

Modify coco.py to fit the current pipeline. Add training and evaluation pipeline for COCO-cat and Synthetic coco-cat dataset (USim and local generated data). Add a notebook that can do statistics for model performance.

# Pull Request Check List

- [x] Updated **documentation** for changed code.

### CODE REVIEW

File: datasetinsights/constants.py [hunk 1]
```diff
@@ -36,7 +36,7 @@
 #   ...
 DEFAULT_DATA_ROOT = "/data"
 SYNTHETIC_SUBFOLDER = "synthetic"
-DEFAULT_PUBLIC_DATASET = "synthetic"
+DEFAULT_PUBLICDEFAULT_KFP_LOG_DIR_DATASET = "synthetic"
```
Comment (adason) [inline datasetinsights/constants.py]: It's not clear where this constant is used. Also, the original constant variable `DEFAULT_PUBLIC_DATASET` is very confusing. Can it this be safely removed? Why is this change required?

File: datasetinsights/configs/faster_rcnn_coco_cat.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+estimator: FasterRCNN
```
Comment (adason) [inline datasetinsights/configs/faster_rcnn_coco_cat.yaml]: Let's not add this single-class yaml file to the repo.

File: datasetinsights/configs/faster_rcnn_coco_cat_synthetic.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+estimator: FasterRCNN
```
Comment (adason) [inline datasetinsights/configs/faster_rcnn_coco_cat_synthetic.yaml]: same here. Let's not add this single class yaml file to repo.

File: kubeflow/compiled/train_and_eval_on_coco_cat.yaml [hunk 4]
```diff
@@ -0,0 +1,258 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (adason) [inline kubeflow/compiled/train_and_eval_on_coco_cat.yaml]: Let's not add single-class object detection pipeline to datasetinsights repo.

File: kubeflow/compiled/train_and_eval_on_coco_cat_synthetic.yaml [hunk 5]
```diff
@@ -0,0 +1,259 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (adason) [inline kubeflow/compiled/train_and_eval_on_coco_cat_synthetic.yaml]: Let's not add single-class object detection pipeline to datasetinsights repo.

File: kubeflow/compiled/train_on_usim_coco_synthetic_dataset.yaml [hunk 6]
```diff
@@ -0,0 +1,269 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (adason) [inline kubeflow/compiled/train_on_usim_coco_synthetic_dataset.yaml]: Let's not add single-class object detection pipeline to datasetinsights repo.

File: kubeflow/pipelines.py [hunk 7]
```diff
@@ -529,3 +529,222 @@ def train_and_evaluate_combined_pipeline(
         num_gpu=evaluate_num_gpu,
         gpu_type=gpu_type,
     )
+
+
+@dsl.pipeline(
```
Comment (adason) [inline kubeflow/pipelines.py]: Let's not add single-class object detection pipeline to datasetinsights repo.

File: notebooks/Model_Performance_Statistics.ipynb [hunk 8]
```diff
@@ -0,0 +1,155 @@
+{
```
Comment (adason) [inline notebooks/Model_Performance_Statistics.ipynb]: This notebook is specific to coco-cats experiment. Let's wait until we have general solution that allow us to evaluate repeated model training + evaluation runs.

**Review by adason — CHANGES_REQUESTED** [review]

File: datasetinsights/constants.py (hunk 1 above)
Comment (BlairLee) [reply datasetinsights/constants.py]: Don't think we use this constant. Will remove it.

File: datasetinsights/datasets/coco.py [hunk 9]
```diff
@@ -143,6 +173,87 @@ def _get_local_annotations_zip(self):
...
+
+            Return:
+                Tuple: (unarchived img path, unarchived annotation path)
+            """
+
+        archive_img_file = Path(data_path) / f"{split}2017.zip"
+        archive_ann_file = Path(data_path) / "annotations_trainval2017.zip"
+        if archive_img_file.exists() and archive_ann_file.exists():
+            img_file_descriptor = os.open(archive_img_file, os.O_RDONLY)
+            try:
```
Comment (adason) [inline datasetinsights/datasets/coco.py]: Violate DRY. It seems line 208-221 is. repeated in 224-238. Can you wrap them in a helper method?

**Review by adason — CHANGES_REQUESTED** [review]

File: datasetinsights/datasets/coco.py (hunk 9 above)
Comment (BlairLee) [reply datasetinsights/datasets/coco.py]: Good call. Wrap it into a function. And test it in [kubeflow](https://datasetinsights.endpoints.unity-ai-thea-test.cloud.goog/_/pipeline/#/runs/details/5fd4ad7a-0718-4a23-a034-395800c46569).

File: datasetinsights/datasets/coco.py [hunk 10]
```diff
@@ -143,6 +173,76 @@ def _get_local_annotations_zip(self):
...
+            # This is for dataset generated by unity simulation.
+            return data_path
+        else:
+            raise DatasetNotFoundError(
+                f"Expecting a file {archive_img_file} and {archive_ann_file}"
+                "under {data_path}"
+            )
+
+    @staticmethod
+    def unarchive_data(data_path, archive_file):
```
Comment (adason) [inline datasetinsights/datasets/coco.py]: Making this private method?

File: datasetinsights/datasets/coco.py [hunk 11]
```diff
@@ -143,6 +173,76 @@ def _get_local_annotations_zip(self):
...
+                shutil.unpack_archive(
+                    filename=archive_file, extract_dir=unarchived_path,
+                )
+                logger.info(f"Unpack {archive_file} to {unarchived_path}")
+        finally:
+            os.close(file_descriptor)
+        return unarchived_path
+
+    @staticmethod
+    def is_dataset_files_present(data_path):
```
Comment (adason) [inline datasetinsights/datasets/coco.py]: private method?

**Comment by BlairLee** [comment]
@adason Is it good to be merged?

File: datasetinsights/datasets/coco.py [hunk 12]
```diff
@@ -143,6 +173,90 @@ def _get_local_annotations_zip(self):
...
+        with open(ann_file_name, "r") as ann_file:
+            anns = json.load(ann_file)
+            for cat in anns["categories"]:
+                label_mappings[cat["id"]] = cat["name"]
+        return label_mappings
+
+    @staticmethod
+    def _preprocess_dataset(data_path, split):
+        """ Preprocess dataset inside data_path and un-archive if necessary.
+
```
Comment (86sanj) [inline datasetinsights/datasets/coco.py:195]: Can we have unit test cases for the new added methods?

Comment (BlairLee) [reply datasetinsights/datasets/coco.py:195]: Added.

File: tests/datasets/test_coco.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+from pathlib import Path
+from unittest.mock import patch
+
+from pytest import raises
+
+from datasetinsights.datasets.coco import CocoDetection
+from datasetinsights.datasets.exceptions import DatasetNotFoundError
+
+
+def test__is_dataset_files_present():
```
Comment (86sanj) [inline tests/datasets/test_coco.py:12]: [nit]: Two '_'

File: tests/datasets/test_coco.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        with open(os.path.join(tmp, "coco.json"), "x"):
+            with open(os.path.join(tmp, "coco.jpg"), "x"):
+                assert CocoDetection._is_dataset_files_present(tmp)
+
+    with tempfile.TemporaryDirectory() as tmp:
+        assert not CocoDetection._is_dataset_files_present(tmp)
+
+
+@patch("datasetinsights.datasets.CocoDetection._unarchive_data")
+def test__preprocess_dataset(mock_unarchive):
```
Comment (86sanj) [inline tests/datasets/test_coco.py:23]: [nit]: Two '_'.

Comment (BlairLee) [reply tests/datasets/test_coco.py:23]: Since it's a private method (original name is `_preprocess_dataset`), so it should be two '_'

Comment (86sanj) [reply tests/datasets/test_coco.py:23]: Is this the naming convention for private method unit test?

Comment (BlairLee) [reply tests/datasets/test_coco.py:23]: If the method name starts with '_', we should include the whole method name in a unit test. For example, `test__add_single_bbox_on_image_crop_label` in `test_visual.py`

Comment (BlairLee) [reply tests/datasets/test_coco.py:23]: Otherwise, we may confuse people if we have another method named 'preprocess_dataset'.

**Review by 86sanj — APPROVED** [review]

**Review by adason — APPROVED** [review]
